// @sanity/google-maps-input/lib/input/GeopointInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _uniqueId2 = _interopRequireDefault(require("lodash/uniqueId"));

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

var _icons = require("@sanity/icons");

var _googleMapsInput = _interopRequireDefault(require("config:@sanity/google-maps-input"));

var _components = require("@sanity/base/components");

var _patchEvent = require("part:@sanity/form-builder/patch-event");

var _changeIndicators = require("@sanity/base/change-indicators");

var _GoogleMapsLoadProxy = require("../loader/GoogleMapsLoadProxy");

var _GeopointSelect = require("./GeopointSelect");

var _GeopointInput = require("./GeopointInput.styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var getStaticImageUrl = value => {
  var loc = "".concat(value.lat, ",").concat(value.lng);
  var params = {
    key: _googleMapsInput.default.apiKey,
    center: loc,
    markers: loc,
    zoom: 13,
    scale: 2,
    size: '640x300'
  };
  var qs = Object.keys(params).reduce((res, param) => {
    return res.concat("".concat(param, "=").concat(encodeURIComponent(params[param])));
  }, []);
  return "https://maps.googleapis.com/maps/api/staticmap?".concat(qs.join('&'));
};

class GeopointInput extends _react.default.PureComponent {
  constructor(props) {
    super(props);

    _defineProperty(this, "_geopointInputId", (0, _uniqueId2.default)('GeopointInput'));

    _defineProperty(this, "editButton", void 0);

    _defineProperty(this, "setEditButton", el => {
      this.editButton = el;
    });

    _defineProperty(this, "handleFocus", event => {
      this.props.onFocus(event);
    });

    _defineProperty(this, "handleBlur", () => {
      this.props.onBlur();
    });

    _defineProperty(this, "handleToggleModal", () => {
      var _this$props = this.props,
          onFocus = _this$props.onFocus,
          onBlur = _this$props.onBlur;
      this.setState(prevState => ({
        modalOpen: !prevState.modalOpen
      }), () => {
        if (this.state.modalOpen) {
          onFocus(['$']);
        } else {
          onBlur();
        }
      });
    });

    _defineProperty(this, "handleCloseModal", () => {
      this.setState({
        modalOpen: false
      });
    });

    _defineProperty(this, "handleChange", latLng => {
      var _this$props2 = this.props,
          type = _this$props2.type,
          onChange = _this$props2.onChange;
      onChange(_patchEvent.PatchEvent.from([(0, _patchEvent.setIfMissing)({
        _type: type.name
      }), (0, _patchEvent.set)(latLng.lat(), ['lat']), (0, _patchEvent.set)(latLng.lng(), ['lng'])]));
    });

    _defineProperty(this, "handleClear", () => {
      var onChange = this.props.onChange;
      onChange(_patchEvent.PatchEvent.from((0, _patchEvent.unset)()));
    });

    this.state = {
      modalOpen: false
    };
  }

  focus() {
    if (this.editButton) {
      this.editButton.focus();
    }
  }

  render() {
    var _this$props3 = this.props,
        value = _this$props3.value,
        compareValue = _this$props3.compareValue,
        readOnly = _this$props3.readOnly,
        type = _this$props3.type,
        markers = _this$props3.markers,
        level = _this$props3.level,
        presence = _this$props3.presence;
    var modalOpen = this.state.modalOpen;

    if (!_googleMapsInput.default || !_googleMapsInput.default.apiKey) {
      return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("p", null, "The ", /*#__PURE__*/_react.default.createElement("a", {
        href: "https://sanity.io/docs/schema-types/geopoint-type"
      }, "Geopoint type"), " needs a Google Maps API key with access to:"), /*#__PURE__*/_react.default.createElement("ul", null, /*#__PURE__*/_react.default.createElement("li", null, "Google Maps JavaScript API"), /*#__PURE__*/_react.default.createElement("li", null, "Google Places API Web Service"), /*#__PURE__*/_react.default.createElement("li", null, "Google Static Maps API")), /*#__PURE__*/_react.default.createElement("p", null, "Please enter the API key with access to these services in", /*#__PURE__*/_react.default.createElement("code", {
        style: {
          whiteSpace: 'nowrap'
        }
      }, "`<project-root>/config/@sanity/google-maps-input.json`")));
    }

    return /*#__PURE__*/_react.default.createElement(_components.FormFieldSet, {
      level: level,
      title: type.title,
      description: type.description,
      onFocus: this.handleFocus,
      onBlur: this.handleBlur,
      __unstable_presence: presence,
      __unstable_changeIndicator: false,
      __unstable_markers: markers
    }, /*#__PURE__*/_react.default.createElement("div", null, value && /*#__PURE__*/_react.default.createElement(_changeIndicators.ChangeIndicatorCompareValueProvider, {
      value: value,
      compareValue: compareValue
    }, /*#__PURE__*/_react.default.createElement(_changeIndicators.ChangeIndicator, {
      compareDeep: true
    }, /*#__PURE__*/_react.default.createElement(_GeopointInput.PreviewImage, {
      src: getStaticImageUrl(value),
      alt: "Map location"
    }))), !readOnly && /*#__PURE__*/_react.default.createElement(_ui.Box, {
      marginTop: 4
    }, /*#__PURE__*/_react.default.createElement(_ui.Grid, {
      columns: 2,
      gap: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Button, {
      mode: "ghost",
      icon: value && _icons.EditIcon,
      padding: 3,
      ref: this.setEditButton,
      text: value ? 'Edit' : 'Set location',
      onClick: this.handleToggleModal
    }), value && /*#__PURE__*/_react.default.createElement(_ui.Button, {
      tone: "critical",
      icon: _icons.TrashIcon,
      padding: 3,
      mode: "ghost",
      text: 'Remove',
      onClick: this.handleClear
    }))), modalOpen && /*#__PURE__*/_react.default.createElement(_ui.Dialog, {
      id: "".concat(this._geopointInputId, "_dialog"),
      onClose: this.handleCloseModal,
      header: "Place the marker on the map",
      width: 1
    }, /*#__PURE__*/_react.default.createElement(_GeopointInput.DialogInnerContainer, null, /*#__PURE__*/_react.default.createElement(_GoogleMapsLoadProxy.GoogleMapsLoadProxy, null, api => /*#__PURE__*/_react.default.createElement(_GeopointSelect.GeopointSelect, {
      api: api,
      value: value,
      onChange: readOnly ? undefined : this.handleChange,
      defaultLocation: _googleMapsInput.default.defaultLocation,
      defaultZoom: _googleMapsInput.default.defaultZoom
    }))))));
  }

}

_defineProperty(GeopointInput, "defaultProps", {
  markers: []
});

var _default = GeopointInput;
exports.default = _default;
