// @sanity/google-maps-input/lib/loader/GoogleMapsLoadProxy.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GoogleMapsLoadProxy = void 0;

var _react = _interopRequireDefault(require("react"));

var _loadGoogleMapsApi = require("./loadGoogleMapsApi");

var _LoadError = require("./LoadError");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class GoogleMapsLoadProxy extends _react.default.Component {
  constructor(props) {
    super(props);

    _defineProperty(this, "loadSubscription", void 0);

    this.state = {
      loadState: 'loading'
    };
    var sync = true;
    this.loadSubscription = (0, _loadGoogleMapsApi.loadGoogleMapsApi)().subscribe(loadState => {
      if (sync) {
        this.state = loadState;
      } else {
        this.setState(loadState);
      }
    });
    sync = false;
  }

  componentWillUnmount() {
    if (this.loadSubscription) {
      this.loadSubscription.unsubscribe();
    }
  }

  render() {
    switch (this.state.loadState) {
      case 'loadError':
        return /*#__PURE__*/_react.default.createElement(_LoadError.LoadError, {
          error: this.state.error,
          isAuthError: false
        });

      case 'authError':
        return /*#__PURE__*/_react.default.createElement(_LoadError.LoadError, {
          isAuthError: true
        });

      case 'loading':
        return /*#__PURE__*/_react.default.createElement("div", null, "Loading Google Maps API");

      case 'loaded':
        return this.props.children(this.state.api) || null;

      default:
        return null;
    }
  }

}

exports.GoogleMapsLoadProxy = GoogleMapsLoadProxy;
