// @sanity/google-maps-input/lib/loader/LoadError.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LoadError = LoadError;

var React = _interopRequireWildcard(require("react"));

var _ui = require("@sanity/ui");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function LoadError(props) {
  var _props$error;

  return /*#__PURE__*/React.createElement(_ui.Card, {
    tone: "critical",
    radius: 1
  }, /*#__PURE__*/React.createElement(_ui.Box, {
    as: "header",
    paddingX: 4,
    paddingTop: 4,
    paddingBottom: 1
  }, /*#__PURE__*/React.createElement(_ui.Text, {
    as: "h2",
    weight: "bold"
  }, "Google Maps failed to load")), /*#__PURE__*/React.createElement(_ui.Box, {
    paddingX: 4,
    paddingTop: 4,
    paddingBottom: 1
  }, props.isAuthError ? /*#__PURE__*/React.createElement(AuthError, null) : /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(_ui.Text, {
    as: "h3"
  }, "Error details:"), /*#__PURE__*/React.createElement("pre", null, /*#__PURE__*/React.createElement(_ui.Code, {
    size: 1
  }, (_props$error = props.error) === null || _props$error === void 0 ? void 0 : _props$error.message)))));
}

function AuthError() {
  return /*#__PURE__*/React.createElement(_ui.Text, null, /*#__PURE__*/React.createElement("p", null, "The error appears to be related to authentication"), /*#__PURE__*/React.createElement("p", null, "Common causes include:"), /*#__PURE__*/React.createElement("ul", null, /*#__PURE__*/React.createElement("li", null, "Incorrect API key"), /*#__PURE__*/React.createElement("li", null, "Referer not allowed"), /*#__PURE__*/React.createElement("li", null, "Missing authentication scope")), /*#__PURE__*/React.createElement("p", null, "Check the browser developer tools for more information."));
}
